### ISSUE TITLE
Correct use of `partition` and `combine`

### ISSUE THREAD

**Issue by metric-space** [patrick-kidger/equinox#1032]
Hello hello

Pretty awesome work with equinox. It's my goto framework of choice for anything deep-learning.

My recent endeavor has me porting over pytorch vgg19 pretrained weights to a vgg19 implementation in  jax & equinox

I was wondering if the use  of `partition` and `combine` here to get over `None`s in the pytree associated with the nonlinearities makes sense ?

FWIW the code does work (checked with inference post-transfer). I'm wondering the current solution is more complex than it needs to be

https://github.com/metric-space/vgg-torch-to-jax-conversion/blob/ef5ca4184399cfa9b9c38d4663a956578fdd748f/vgg_conversion.py?plain=1#L125-L154

```python

    key = jax.random.PRNGKey(2345)

    vgg19 = models.vgg19(pretrained=True)  # from pytorch

    jax_model = VGG19(key) # implemented in jax & equinox

    model_tree_to_replace, model_tree_not_to_replace = eqx.partition(
        jax_model, eqx.is_array
    )

    leaves, treedef = jax.tree_util.tree_flatten(model_tree_to_replace)

    pt_params = []
    for i, (k, v) in enumerate(vgg19.state_dict().items()):
        arr = v.numpy()
        if arr.ndim != leaves[i].ndim:
            arr = jnp.array(arr)[:, None, None]
        else:
            arr = jnp.array(arr)
        pt_params.append(arr)

    revitalized_jax_model = treedef.unflatten(pt_params)
    revitalized_jax_model = eqx.combine(
        revitalized_jax_model, model_tree_not_to_replace
    )


```

**Comment by patrick-kidger** [comment]
I think this looks not-too-crazy 😁 
If you want an alternative then I think this might feel slightly safer, as it doesn't rely on the order of the weights in the state dictionary matching the order of the flattened Equinox model: https://github.com/patrick-kidger/equinox/issues/657#issuecomment-1943786874

**Comment by metric-space** [comment]
Thank you for the answer
